xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWiki.java

```diff
@@ -2435,17 +2453,61 @@ public String getSkinFile(String filename, boolean forceSkinAction, XWikiContext
 
         // If all else fails, use the default base skin, even if the URLs could be invalid.
         URL url;
+
         if (forceSkinAction) {
             url = urlf.createSkinURL(filename, "skins", getDefaultBaseSkin(context), context);
         } else {
             url = urlf.createSkinURL(filename, getDefaultBaseSkin(context), context);
         }
-        return urlf.getURL(url, context);
+        queryParameters.putAll(getResourceURLParameters(url));
+        return urlf.getURL(url, context, queryParameters);
+    }
+
+    public String getSkinFile(String filename, boolean forceSkinAction, XWikiContext context)
+    {
+        return getSkinFile(filename, forceSkinAction, context, new LinkedHashMap<>());
+    }
+
+    private Map<String, String> getResourceURLParameters(Resource<?> resource)
+    {
+        try {
+            URL resourceUrl = getResource(resource.getPath());
+            return getResourceURLParameters(resourceUrl);
+        } catch (MalformedURLException e) {
+            LOGGER.debug("Error while getting URL for resource filepath [{}]", resource.getPath(), e);
+            return Collections.singletonMap(CACHE_VERSION, getVersion());
+        }
+    }
+
+    private Map<String, String> getResourceURLParameters(String resourceFilePath)
+    {
+        try {
+            URL resourceUrl = getResource(resourceFilePath);
+            return getResourceURLParameters(resourceUrl);
+        } catch (MalformedURLException e) {
+            LOGGER.debug("Error while getting URL for resource filepath [{}]", resourceFilePath, e);
+            return Collections.singletonMap(CACHE_VERSION, getVersion());
+        }
+    }
+
+    private Map<String, String> getResourceURLParameters(URL resourceUrl)
+    {
+        Map<String, String> parameters = new LinkedHashMap<>();
+
+        try {
+            Path resourcePath = Paths.get(resourceUrl.toURI());
+            FileTime lastModifiedTime = Files.getLastModifiedTime(resourcePath);
+            parameters.put(CACHE_VERSION, String.valueOf(lastModifiedTime.toMillis()));
```

Comment:
For resources that have an associated last modification date I think the best practice is to use https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since .

Reply:
I don't think it applies on the current workflow. If I understand well `If-Modified-Since` it's a request header so it's when the browser needs to load a resource that it should fill this header. Now here we only send an URL to the browser to make the request. I don't see how/where we could indicate such a date to put in the right header.

Reply:
My suggestion was to fix the code that serves the resource (e.g. Skin action) instead of the code that builds the resource URL. The code that serves the resource can set and read HTTP headers.

Reply:
OK but then it's the `Last-Modified` header which should be changed: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified as a response header.

Reply:
Yes, the WebJars resource handler does something like this:
* sets the Last-Modified header https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-resource/xwiki-platform-resource-servlet/src/main/java/org/xwiki/resource/servlet/AbstractServletResourceReferenceHandler.java#L216
* and then checks the If-Modified-Since header https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-resource/xwiki-platform-resource-servlet/src/main/java/org/xwiki/resource/servlet/AbstractServletResourceReferenceHandler.java#L114

Maybe something like this can be applied to skin resources.

Reply:
interesting, I'll check that

Reply:
So I indeed check what you did, but as far as I can see it's difficult to apply right now since we don't use a ResourceReferenceHandler for the bin (see the comment in EntityResourceReferenceHandler https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-resource/xwiki-platform-resource-default/src/main/java/org/xwiki/resource/internal/entity/EntityResourceReferenceHandler.java#L55-L58) 
I might be wrong here but it seems to me that we should first fix that in order to properly apply the headers.
Since this looks like a big refactoring I propose that we apply first the easy solution already implemented with the query parameters.

Reply:
I wasn't suggesting you to use exactly the code from ``AbstractServletResourceReferenceHandler``. I linked it just as an example. My point was actually to use the **cache-related HTTP headers** instead of the query string. I find it wrong to put the date in the query string if there are dedicated HTTP headers to handle this.